Web frontend attack support

awwad opened the pull request:
Add high-level functions that execute new attacks in the demonstration and make them available via XML-RPC to the web frontend.

[Edited by @vladimir.v.diaz to list the attacks added in this pull request]

Attacks added to Director repository:
```
- prepare_replay_attack_nokeys
- replay_attack_nokeys
- undo_replay_attack_nokeys

- keyed_arbitrary_package_attack
- undo_keyed_arbitrary_package_attack
```

Attacks added to Image repository:
```
- keyed_arbitrary_package_attack
- undo_keyed_arbitrary_package_attack
```

awwad committed:
Add functions web frontend can call to attack Director

These execute all the commands necessary at each stage of the
attack demonstration, so that buttons can be created that
trigger the attacks.

- prepare_replay_attack_nokeys
- replay_attack_nokeys
- undo_replay_attack_nokeys
- keyed_arbitrary_package_attack
- undo_keyed_arbitrary_package_attack

awwad committed:
Add functions web frontend can call to attack Image Repo

These execute all the commands necessary at each stage of the
attack demonstration, so that buttons can be created that
trigger the attacks.

- keyed_arbitrary_package_attack
- undo_keyed_arbitrary_package_attack

awwad committed:
Correct typo in print statement in demo Image Repo

awwad committed:
Register new attack functions via XML-RPC for web frontend

awwad commented:
Fixes #68

vladimir-v-diaz commented on demo/demo_director.py:
```diff
@@ -742,8 +742,30 @@ def listen():
+      'undo_mitm_arbitrary_package_attack')
+
+  # Attack 2: Replay, no keys
+  server.register_function(prepare_replay_attack_nokeys,
```
In the demo instructions, we use rollback instead of replay.  We should try to use consistent names.

vladimir-v-diaz replied:
The Uptane Design Overview doc uses Rollback Attack, so we should make sure the code uses that name.

awwad replied:
This is not a rollback attack, though. We're replaying old metadata which, even if it were accepted, would not move the client to a lower version of an image.

awwad replied:
The definition of rollback provided in the design doc is:
* *Figure 7.3.3a: Rollback attack, where attackers cause ECUs to install obsolete software. The vehicle has version 10 of ECU-1 software, but the attackers send version 1 of the software as the latest update, even though version 12 exists on the repository.*

This attack is replaying valid but outdated metadata.

I'm going to switch the other references to 'replay' from 'rollback' as well.

vladimir-v-diaz replied:
Is this a Freeze attack, then?

awwad replied:
I guess so. Breaking it down:

The attack *here* only replays an old timestamp, and doesn't result in a firmware rollback. So _this attack_ is:
- a [Replay attack](https://en.wikipedia.org/wiki/Replay_attack) by common definition: because it replays previously used, otherwise valid information
- a [Rollback attack according to the TUF Security Doc](https://github.com/theupdateframework/tuf/blob/develop/SECURITY.md#attacks-and-weaknesses), which has a very general definition ("older **files** than those the client has already seen")
- NOT a [Rollback attack according to the TUF Spec itself](https://github.com/theupdateframework/tuf/blob/40aaf93f5a39f659cc312b66624f2cbff59b8582/docs/tuf-spec.txt#L166), because that refers to installing older software, which this does not do
- NOT a [Rollback attack according to the Uptane Design Document](https://docs.google.com/document/d/1pBK--40BCg_ofww4GES0weYFB6tZRedAjUy6PJ4Rgzk/edit#heading=h.4mo91b3dvcqd), which also defines rollback as installation of older software than the vehicle has.

Similar questions arise for the freeze attack label.

We don't really _have_ to address this (though it'd be nice to have a consistent definition). We could expand the attack here to perform a rollback of an installed image, with a little bit of effort.

awwad replied:
I pushed some clarifications for now that should suffice. If someone wants to address these definitions later, or if we want to expand the replay attack into a rollback attack, we can still do that later.

lukpueh commented on demo/demo_image_repo.py:
```diff
@@ -251,7 +251,7 @@ def listen():
 
   if xmlrpc_service_thread is not None:
-    print('Sorry - there is already a Director service thread listening.')
+    print('Sorry: there is already a listening Image Repository service thread')
```
I'd stick with one delimiter... 
```shell
$ grep -r "there is already" .
./demo/demo_director.py:    print('Sorry, there is already a server process running.')
./demo/demo_director.py:    print('Sorry - there is already a Director service thread listening.')
./demo/demo_image_repo.py:    print('Sorry, there is already a server process running.')
./demo/demo_image_repo.py:    print('Sorry: there is already a listening Image Repository service thread')
```
Really not a blocker though!

lukpueh commented on demo/demo_director.py:
```diff
@@ -742,8 +742,30 @@ def listen():
+  server.register_function(sign_with_compromised_keys_attack,
+      'sign_with_compromised_keys_attack')
+  server.register_function(undo_sign_with_compromised_keys_attack,
+      'undo_sign_with_compromised_keys_attack')
```
Do Attack 1 - 4 refer to [Attack 3.1 - 3.4 in updaten/README.md](https://github.com/uptane/uptane/blame/develop/README.md#L245-L377)? I.e.:
```
"1: Arbitrary Package, no keys" --> "3.1: Arbitrary Package Attack on Director Repository without Compromised Keys"
"2: Replay, no keys" --> "3.2: Arbitrary Package Attack on Image Repository without Compromised Keys"
"3: Arbitrary Package, keyed" --> 3.3: Rollback Attack without Compromised Keys
"4: Arbitrary Package, Revoked Key" --> "3.4: Arbitrary Package Attack with a Compromised Director Key"
```
If so, I'd just copy paste the headings from README.md...

lukpueh replied:
Same goes for `demo_image_repo.py`.

awwad replied:
Addressed.

vladimir-v-diaz commented on demo/demo_director.py:
```diff
@@ -936,6 +958,70 @@ def restore_timestamp(vin):
+def undo_keyed_arbitrary_package_attack(vin, ecu_serial, target_filepath):
+
+  # Revoke potentially compromised keys, replacing them with new keys.
+  revoke_and_add_new_keys_and_write_to_live()
```
This function has been renamed to `revoke_compromised_keys()`.   Has this PR been sync'd with the latest code on master?

awwad replied:
addressed

vladimir-v-diaz commented on demo/demo_image_repo.py:
```diff
@@ -263,12 +263,32 @@ def listen():
   # frontend.
-  server.register_function(add_target_to_imagerepo, 'add_target_to_supplier_repo')
+  server.register_function(add_target_to_imagerepo,
+      'add_target_to_supplier_repo')
```
`supplier` is used here.  Is there a reason why it cannot be updated to use imagerepo?

awwad replied:
No, just have to do it together with the web frontend changes.

lukpueh commented on demo/demo_image_repo.py:
```diff
@@ -263,12 +263,32 @@ def listen():
   # frontend.
-  server.register_function(add_target_to_imagerepo, 'add_target_to_supplier_repo')
+  server.register_function(add_target_to_imagerepo,
+      'add_target_to_supplier_repo')
```
Is there a good reason to have different names for the function and the XMLRPC interface? Btw. for all other calls to `server.register_function`, which do use the same name for both, the second [argument is optional](https://docs.python.org/2/library/simplexmlrpcserver.html#SimpleXMLRPCServer.SimpleXMLRPCServer.register_function).

lukpueh replied:
Maybe (that's how we do this on the clearinghouse) you could prefix functions with `xmlrpc_` but expose the name without the prefix to the Interface.

awwad replied:
No good reason anymore, no. Will resolve in parallel with changes to web frontend.

vladimir-v-diaz commented:
See my in-line review comments.

Note: All of the functions added are missing docstrings.

lukpueh commented on demo/demo_director.py:
```diff
@@ -936,6 +958,70 @@ def restore_timestamp(vin):
+  rollback attack with no compromised keys against the Director.
+
+  After this is done, the Primary should update, then the replay attack function
+  should be run (replay_attack_nokeys).
```
I think it would be a good idea to mention what the preparation actually is/does -- `backup_timestamp`, `write_to_live`?

awwad replied:
Addressed.

lukpueh commented on demo/demo_director.py:
```diff
@@ -936,6 +958,70 @@ def restore_timestamp(vin):
+
+
+
+
```
In general, the functions exposed via XMLRPC should follow the [extended docstring code style guidelines](https://github.com/secure-systems-lab/code-style-guidelines#documentation-strings), i.e. purpose, arguments, ...

lukpueh replied:
Same goes for `demo_image_repo.py`

awwad replied:
Fair enough, yes. But I'm going to merge this now that other comments are addressed, and I'll expand the docstrings further in the future.

... after I run the attacks again, that is.

lukpueh reviewed (commented):
LGTM, my comments are mostly related to naming and commenting.

lukpueh reviewed (commented):
*(accidental duplicate of my review summary)*

awwad committed:
Update function name for demo key revocation

from revoke_and_add_new_keys_and_write_to_live to
revoke_compromised_keys.

That change was made in another PR that was merged,
so it must be propagated here.

awwad committed:
Clarify replay attacks in README and demo Director

connecting them to rollback attacks in the design document.

awwad committed:
DOC: Add docstrings to demo Director and ImageRepo functions

especially those relating to attack calls for the web frontend.

awwad committed:
minor: Consistent delimiters in print statements :P

awwad committed:
DOC: Document attacks in demo Director and Image Repo

in response to review comments by @lukpueh

awwad committed:
Remove XMLRPC registration for nonexistent demo attack on ImageRepo

We aren't using an arbitrary package attack using an already-revoked
key on the Image Repo; that attack against the Director repo
suffices.

awwad committed:
bugfix demo replay attack from merging

On another, merged branch, rollback was renamed to replay.
Code added on this PR's branch must also use 'replay'.

awwad committed:
DOC: Convert some comments to demo docstrings

awwad committed:
Add print statements to demo attacks' start and finish

for demo Director and demo Image Repo, and also add some spacing.

awwad committed:
Fix web frontend's version of the arbitrary package attack w/ key

by checking for the right file location. It was previously checking in
the root directory rather than in the repository's directory.

This was done for both demo Director and demo Image Repo.

Also add some clarifying comments.

awwad committed:
Fix bug from bad print statement and and missing print in demo

in demo Image Repo.

awwad committed:
Fix bug: 'dd' reference in demo Director and Image Repo

was breaking web frontend's version of keyed arbitrary package attack.

Also fix minor typo in print statement.

awwad committed:
Fix typo in demo Image Repo attack print statement
